### ISSUE TITLE
the train part

### ISSUE THREAD

**Issue by zzb112233** [brijeshiitg/Pytorch-implementation-of-SRNet#5]
Dear author,can you provide the code of the training part? Or Is there anyone who has developed the train part successfully? Reaaaaaly appreciate!!

**Comment by yiwliu** [comment]
> Dear author,can you provide the code of the training part? Or Is there anyone who has developed the train part successfully? Reaaaaaly appreciate!!

hey friends,  http://dde.binghamton.edu/download/feature_extractors/ <-this link has the SRNet (tensorflow type) can you try to train and display the result?
I had trained the tensorflow type SRNet, But I didn't get good result. I think my datasets wrong. I use BOSSbase_1.01 60% as train cover, 10% as valid cover. and generate corresponding stego with s-uniward(0.4)

**Comment by zzb112233** [comment]
> > Dear author,can you provide the code of the training part? Or Is there anyone who has developed the train part successfully? Reaaaaaly appreciate!!
> 
> hey friends, http://dde.binghamton.edu/download/feature_extractors/ <-this link has the SRNet (tensorflow type) can you try to train and display the result?
> I had trained the tensorflow type SRNet, But I didn't get good result. I think my datasets wrong. I use BOSSbase_1.01 60% as train cover, 10% as valid cover. and generate corresponding stego with s-uniward(0.4)

I have trained the tf type. I use BOSSbase_1.01 too,the devide setting is 5000:1000:4000,the acc of s-uniward(0.4) I got is 0.74.The image size I use is 256*256,and I embedded it myself instead of using the dataset Jessica has improve(when I use hers,My network can`t train,I dont`t know why)

**Comment by pixml27** [comment]
Hello! 
I also saw this link, but did not want to use this code due to the fact that it is written in the python 2
I tried to train this pytorch implementation, but my accuracy and roc-auc is 0.5 after 50 epochs. So, this my part of code, maybe we we can work together to develop the best learning strategy
I also saw an issue here whether to use normalization - I don't know for sure, I didn't find anything about it in the original article ... Also  interesting question are how to split data for train/test. I have 30k images for training (15k cover and 15k corresponding them JUNIWARD), and 10 k for validation (5k 5 Jun), i choosed alaska2 dataset

```
class StegoDataset(Dataset):
    def __init__(self, path, df, transforms):
        self.path = path
        self.df = df
        self.transforms = transforms

    def __getitem__(self, item):
        path = os.path.join(self.path, f"{self.df.newpath.iloc[item]}")
        
        label = torch.Tensor([self.df.lab.iloc[item]])
        img = cv2.imread(path)
        img = self.transforms(image=img)['image']
        img = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
        img = torch.from_numpy(np.array([img]))
        return img.permute(0, 1,2), label
    
    def __len__(self):
        return len(self.df)
    
model=Srnet()
model.cuda()
optimizer = Adamax(model.parameters())
criterion = nn.CrossEntropyLoss()

train_ds = StegoDataset(
    "datasets/stego/alldata/", 
    train_df, 
    A.Compose([
        #A.Normalize(),
        A.Resize(256,256)
    ]),
)
train_loader = DataLoader(train_ds, 16, shuffle=True)

valid_ds = StegoDataset(
    "datasets/stego/alldata/", 
    test_df, 
    A.Compose([
        #A.Normalize(),
        A.Resize(256,256)
    ]),
)

 valid_loader = DataLoader(valid_ds, 16, shuffle=False)
 
```
test_df and train_df - dataframes, where are path column and lab (labels) column


```
def train_epoch(num, loader):
    st=time.time()
    model.train()
    running_loss = 0.
    for i, (batch, labels) in enumerate(loader):
        batch = batch.cuda()
        labels = labels.cuda()
        optimizer.zero_grad()
        outputs = model(batch.float())
        loss = criterion(outputs.softmax(dim=1), torch.max(labels, 1)[1])
        loss.backward()
        optimizer.step()

        running_loss += loss.item()
    
    return running_loss / (i+1)
def eval_epoch(num, loader):
    running_loss = 0.
    y,y_score=[],[]
    for i, (batch, labels) in enumerate(loader):
        batch = batch.cuda()
        labels = labels.cuda()
        outputs = model(batch.float())
        
        y_score.append(outputs.softmax(dim=1).cpu().detach())
        y.append(labels.cpu().detach())
    scores=[torch.argmax(item) for sublist in y_score for item in sublist]
    labels=[item for sublist in y for item in sublist]
    return running_loss/(i+1),roc_auc_score(labels, scores),accuracy_score(labels, scores)

def train_and_eval(n):
    for epoch in range(n):
        los=train_epoch(epoch, train_loader)
        ev,roc,acc=eval_epoch(epoch, valid_loader)
        with open("result.csv","a") as f:
            st=str(los)+','+str(ev)+','+str(roc)+','+str(acc)+"\n"
            f.write(st)
            f.close()
```

**Comment by pixml27** [comment]
I have a lot of mistakes in code above. this is the best my implementation of train part
i've just started train srnet and maybe it  would be better than 0.5 auc  - 0.5001,0.5005,0.5008 - rocauc in first 3 epochs
And one important thing! I think one batch (16 images in my case) should contain 8 pairs of cover-stego images, not random images with cover/stego labels

My code:
```
class StegoDataset(Dataset):
    def __init__(self, path, df, transforms):
        self.path = path
        self.df = df
        self.transforms = transforms
        
    def __getitem__(self, item):
        path = os.path.join(self.path, f"{self.df.newpath.iloc[item]}")
        
        label = torch.Tensor([self.df.lab.iloc[item]])
        img = cv2.imread(path)
        img = self.transforms(image=img)['image']
        img = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
        img = img / 255
        img = torch.from_numpy(np.array([img]))
        return img.permute(0, 1,2), label
    
    def __len__(self):
        return len(self.df)
    
model=Srnet()
model.cuda()
optimizer = Adamax(model.parameters())
criterion = nn.CrossEntropyLoss()

train_ds = StegoDataset(
    '/home/akor/datasets/stego/alldata/', 
    train_df, 
    A.Compose([
        #A.Normalize(),
        A.Resize(256,256)
    ]),
)
train_loader = DataLoader(train_ds, 16, shuffle=False)

valid_ds = StegoDataset(
    '/home/akor/datasets/stego/alldata/', 
    test_df, 
    A.Compose([
        #A.Normalize(),
        A.Resize(256,256)
    ]),
)

valid_loader = DataLoader(valid_ds, 16, shuffle=True)
def train_epoch(num, loader):
    st=time.time()
    model.train()
    running_loss = 0.
    for i, (batch, labels) in tqdm.tqdm(enumerate(loader), position=0, leave=True):
        batch = batch.cuda()
        labels=torch.tensor([item for sublist in labels for item in sublist],dtype=torch.long)
        labels = labels.cuda()
        optimizer.zero_grad()
        outputs = model(batch.float())
        loss = criterion(outputs.softmax(dim=1), labels)
        loss.backward()
        optimizer.step()
        running_loss += loss.item()
    return running_loss / (i+1)
def eval_epoch(num, loader):
    running_loss = 0.
    y,y_score=[],[]
    model.eval()
    outputs = []
    loss=[]
    y_prob=[]
    with torch.no_grad():
        for i, (batch, labels) in tqdm.tqdm(enumerate(loader), position=0, leave=True):
            batch = batch.cuda()
            labels=torch.tensor([item for sublist in labels for item in sublist],dtype=torch.long)
            labels = labels.cuda()
            outputs = model(batch.float())
            loss = criterion(outputs.softmax(dim=1), labels)
            running_loss += loss.item()
            y_score.append(outputs.softmax(dim=1).cpu().detach())
            y_prob.append(outputs.softmax(dim=1).cpu().detach()[:,1:])
            y.append(labels.cpu())
        
    scores=[torch.argmax(item) for sublist in y_score for item in sublist]
    labels=[item for sublist in y for item in sublist]
    y_prob=[item for sublist in y_prob for item in sublist]
    return running_loss/(i+1),roc_auc_score(labels,y_prob),accuracy_score(labels, scores)

def train_and_eval(n):
    for epoch in range(n):
        los=train_epoch(epoch, train_loader)
        ev,roc,acc=eval_epoch(epoch, valid_loader)
        with open("result.csv","a") as f:
            st=str(los)+','+str(ev)+','+str(roc)+','+str(acc)+"\n"
            f.write(st)
            f.close()


        
train_and_eval(100)
```

**Comment by brijeshiitg** [comment]
Use data augmentation while training. Without data augmentation, it won't train

**Comment by zzb112233** [comment]
My train acc always comes to 0.69 and it cant descend. It's so hard to me to implement the train part. Can the owner provide the train part code?THANKS A LOT !!!

**Comment by zzb112233** [comment]
Dear author,the amount of epochs of the tensorflow implementation(the paper set)is 457,but according to the result png,your training epoch is only 100.Also,with my train code,it`s always lead to convergence problems.Can you point out what learning rate and what kind of preeprocessing you use?Thanks to response,best wishes!
